src/index.js

```diff
@@ -135,12 +135,21 @@ function matchesSelector (element, ast) {
       if (node.value === ' ') {
         // walk all ancestors
         const precedingNodes = getLastNonCombinatorNodes(nodes.slice(0, i))
-        const ancestor = getFirstMatchingAncestor(element, precedingNodes)
+        let ancestor = getFirstMatchingAncestor(element, precedingNodes)
         if (!ancestor) {
           return false
         } else {
-          element = ancestor
-          i -= precedingNodes.length
+          // Eventhough first ancestor matches the selector, ancestor shall match all preceding nodes
+          while (ancestor) {
+            // If this ancestor is compatible with the preceding nodes, then it is a match
+            // If not, walk up the ancestor tree until a match is found
+            if (matchesSelector(ancestor, { nodes: nodes.slice(0, i - precedingNodes.length) })) {
+              break
+            }
+            ancestor = getFirstMatchingAncestor(ancestor, precedingNodes)
+          }
+          // If ancestor is null/undefined, it means while loop has exhausted all the possible ancestors and not found a match
+          return ancestor != null
         }
```

Comment:
This whole section can be simplified to:

```js
        const precedingNodes = getLastNonCombinatorNodes(nodes.slice(0, i))
        const prePrecedingNodes = nodes.slice(0, i - precedingNodes.length)
        let ancestor = getFirstMatchingAncestor(element, precedingNodes)
        if (!ancestor) {
          return false
        }
        // Even though first ancestor matches the selector, ancestor should match all preceding nodes
        while (ancestor) {
          // If this ancestor is compatible with the preceding nodes, then it is a match
          // If not, walk up the ancestor tree until a match is found
          if (matchesSelector(ancestor, { nodes: prePrecedingNodes })) {
            return true
          }
          ancestor = getFirstMatchingAncestor(ancestor, precedingNodes)
        }
        // While loop has exhausted all the possible ancestors and not found a match
        return false
```

However, I don't actually understand what `prePrecedingNodes` is supposed to be. (Hence the silly name.)

I worry that `prePrecedingNodes` is not actually correct, and is just hyper-optimized for the particular selector you are trying to test.

I also note that we are effectively breaking out of the top-level `for` loop and doing our own matching system where we loop inside of the loop... This seems brittle. Perhaps adding more tests for multiple combinators (e.g. `.foo .bar .baz`, `.foo .bar .baz. quux`) would make the algorithm clearer.

Reply:
I agree that this is not a perfect and clean approach. When there is a space combinator, we shall look at all the parent/ancestor possibilities before returning false. That's the basic idea behind this approach.

I will surely add more test cases as suggested. By the way, I'm using the modified version in one of my wdio project and it's working without any hiccups. 

I found another similar library which is trying to solve the shadow dom problem. Their approach of breaking the selectors and processing them individually is interesting. You can take a look here https://github.com/webdriverio/query-selector-shadow-dom#readme

Reply:
The space combinator is a special case in the CSS selector world.
The library which I mentioned above, is breaking the selectors are the space and processing them separately.
Our algo should also look at doing something along similar lines

```
querySelectorShadowDom.querySelectorDeep("outer-component > div:nth-child(1) .inner-component:nth-child(1) select")
domqueryselector:206 Querying shadowdom 
(3) [Array(2), Array(1), Array(1)]
0
: 
(2) ['outer-component', 'div:nth-child(1)']
1
: 
['.inner-component:nth-child(1)']
2
: 
['select']
```